src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java

```diff
@@ -194,103 +183,134 @@ public VertexConsumer getBuffer(RenderLayer layer) {
-				if (isVisiblePos(rrp, cameraPos)) {
-					// For the current region, rebuild each render layer using the buffer builders
-					// Find all block entities in this region
-					if (currentWorld == null) {
-						break;
-					}
+			Vec3d cam = wrc.camera().getPos();
 
-					for (int chunkX = rrp.x << REGION_FROMCHUNK_SHIFT; chunkX < (rrp.x + 1) << REGION_FROMCHUNK_SHIFT; chunkX++) {
-						for (int chunkZ = rrp.z << REGION_FROMCHUNK_SHIFT; chunkZ < (rrp.z + 1) << REGION_FROMCHUNK_SHIFT; chunkZ++) {
-							blockEntities.addAll(currentWorld.getChunk(chunkX, chunkZ).getBlockEntities().values());
+			if (!needsRebuild.isEmpty()) {
+				wrc.profiler().swap("rebuild");
+				// Make builders for regions that are marked for rebuild
+				for (RenderRegionPos rrp : needsRebuild) {
+					builders.compute(rrp, (k, v) -> {
+						if (v != null) v.reset();
+						else v = new RegionBufferBuilder();
+						return v;
+					});
+				}
+				// Iterate over all RegionBuilders, render and upload to RegionBuffers
+				Set<RenderRegionPos> removing = Sets.newLinkedHashSet();
+				List<BlockEntity> blockEntities = new ArrayList<>();
+				for (Map.Entry<RenderRegionPos, RegionBufferBuilder> entryBuilder : builders.entrySet()) {
+					RenderRegionPos rrp = entryBuilder.getKey();
+					if (isVisiblePos(rrp, cam)) {
+						// For the current region, rebuild each render layer using the buffer builders
+						// Find all block entities in this region
+						if (currentWorld == null) {
+							break;
 						}
-					}
 
-					entryBuilder.getValue().build(blockEntities);
-					blockEntities.clear();
+						for (int chunkX = rrp.x << REGION_FROMCHUNK_SHIFT; chunkX < (rrp.x + 1) << REGION_FROMCHUNK_SHIFT; chunkX++) {
+							for (int chunkZ = rrp.z << REGION_FROMCHUNK_SHIFT; chunkZ < (rrp.z + 1) << REGION_FROMCHUNK_SHIFT; chunkZ++) {
+								blockEntities.addAll(currentWorld.getChunk(chunkX, chunkZ).getBlockEntities().values());
+							}
+						}
+
+						if (!blockEntities.isEmpty()) {
+							entryBuilder.getValue().bake(blockEntities);
+							blockEntities.clear();
+						} else {
+							removing.add(rrp);
+						}
 
-					RegionBuffer buf = regions.computeIfAbsent(entryBuilder.getKey(), k -> new RegionBuffer());
-					for (Map.Entry<RenderLayer, BufferBuilder> layerBuilder : entryBuilder.getValue()) {
-						RenderLayer layer = layerBuilder.getKey();
-						BufferBuilder builder = layerBuilder.getValue();
+						RegionBuffer buf = regions.computeIfAbsent(rrp, k -> new RegionBuffer());
+						for (Map.Entry<RenderLayer, BufferBuilder> layerBuilder : entryBuilder.getValue()) {
+							RenderLayer layer = layerBuilder.getKey();
+							BufferBuilder builder = layerBuilder.getValue();
 
-						if (!builder.isBuilding()) builder.begin(layer.getDrawMode(), layer.getVertexFormat());
-						buf.upload(layer, builder);
+							if (!builder.isBuilding()) builder.begin(layer.getDrawMode(), layer.getVertexFormat());
+							buf.upload(layer, builder);
```

Comment:
Why do you start building only to immediately upload it? The presence of a `isBuilding` guard is a code smell, you should already know the state of your builders...

Reply:
1: I start building it right before uploading so the VBO is empty and does not contain the data from the last rebuild
2: The `isBuilding` guard is needed so when you make a text block empty it doesn't crash

Please correct me if I am wrong with any of these answers

Reply:
Removed the code smell, added proper fixes to what it was trying to fix.